AGENTS.md

```diff
@@ -0,0 +1,187 @@
+
+# Update test metadata (for tests in test/ directory)
+make update
+
+# Verification
+make verify             # Runs verify-yaml and verify-update
+make verify-yaml        # Validates YAML manifests
+make verify-update      # Ensures generated files are up-to-date
+```
+
+## Architecture
+
+### Core Components
+
+#### Main Operator Loop (pkg/cvo/cvo.go)
+- The `Operator` struct is the central controller
+- Watches the `ClusterVersion` resource in cluster
+- Implements a reconciliation loop that is the single writer of `ClusterVersion` status
+- Uses a work queue pattern with 15 max retries for failed operations
+- Periodically checks for updates from configured upstream server
+- Maintains conditions: `Progressing`, `Available`, `Failing`, `Upgradeable`, `RetrievedUpdates`
+
+#### SyncWorker (pkg/cvo/sync_worker.go)
+- Abstracts payload synchronization via `ConfigSyncWorker` interface
+- Manages the actual application of manifests from release payload images
+- Handles payload retrieval, verification, and application
+- Uses rate limiting and retry logic for resilient updates
+- Reports status back to main operator via `SyncWorkerStatus` channel
+
+#### Payload Management (pkg/payload/)
+- `State` enum defines update modes: `UpdatingPayload`, `ReconcilingPayload`, `InitializingPayload`, `PrecreatingPayload`
+- `UpdatingPayload`: Conservative ordering when transitioning versions, errors block dependent operators
+- `ReconcilingPayload`: Maintains current state, recreates resources without strict ordering
+- `InitializingPayload`: First-time deployment, fast progress with tolerance for transient errors
+- Release payloads stored in two directories:
+  - `manifests/`: CVO's own manifests
+  - `release-manifests/`: Manifests for other cluster operators
+- Task graph orchestrates ordered application of manifests based on dependencies
+
+#### Resource Libraries (lib/)
+- `resourcebuilder/`: Builds Kubernetes resources from manifests
+- `resourceapply/`: Applies resources with proper merge semantics for different API types
+- `resourcemerge/`: Merges resource specifications (handles fields like `hostUsers` flag)
+- `resourceread/`: Reads and validates manifests
+- `capability/`: Manages cluster capability filtering
+
+#### Update Service Integration (pkg/cincinnati/)
+- Communicates with OpenShift Update Service (Cincinnati) to fetch available updates
+- Parses update graphs with nodes (release versions) and edges (upgrade paths)
+- Supports conditional edges based on cluster conditions
+
+#### Preconditions (pkg/payload/precondition/)
+- Validates cluster readiness before applying updates
+- Checks prevent upgrades when cluster is in unhealthy state
+- ClusterVersion-specific preconditions in `precondition/clusterversion/`
+
+### Entry Points
+
+#### cmd/cluster-version-operator/main.go
+- Uses cobra for CLI structure
+- Actual start logic delegated to `pkg/start/` package
+
+#### pkg/start/start.go
```

Comment:
Another `#` to nest this under the calling `main.go` entry point?

Reply:
Added.